tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala

```diff
@@ -0,0 +1,594 @@
+              }
+          }
+        }
+
+      case _ =>
+        throw new TypingInputException(s"""Error in ${decl.name}: Expected /\ AssumeType(varName, "typeString") /\ ...""")
+    }
+  }
+
+  // TODO: a long string of translation rules. Can we decompose it?
+  def apply(ex: TlaEx): EtcExpr = ex match {
+    case NameEx(name) =>
+      // x becomes x
+      mkName(ExactRef(ex.ID), name)
+
+    //*********************************************** LITERALS **************************************************
+    case ValEx(TlaInt(_)) =>
+      // an integer literal becomes IntT1
+      mkConst(ExactRef(ex.ID), IntT1())
+
+    case ValEx(TlaBool(_)) =>
+      // a Boolean literal becomes BoolT1
+      mkConst(ExactRef(ex.ID), BoolT1())
+
+    case ValEx(TlaStr(_)) =>
+      // a string literal becomes StrT1
+      mkConst(ExactRef(ex.ID), StrT1())
+
+    case ValEx(TlaReal()) =>
+      // a real literal becomes RealT1
+      mkConst(ExactRef(ex.ID), RealT1())
+
+    case ValEx(TlaIntSet) =>
+      // the set of all integers is SetT1(IntT1)
+      mkConst(ExactRef(ex.ID), SetT1(IntT1()))
+
+    case ValEx(TlaNatSet) =>
+      // the set of all naturals is SetT1(IntT1)
+      mkConst(ExactRef(ex.ID), SetT1(IntT1()))
+
+    case ValEx(TlaRealSet) =>
+      // the set of all reals is SetT1(RealT1)
+      mkConst(ExactRef(ex.ID), SetT1(RealT1()))
+
+    case ValEx(TlaBoolSet) =>
+      // the set of all Booleans is SetT1(BoolT1)
+      mkConst(ExactRef(ex.ID), SetT1(BoolT1()))
+
+    case ValEx(TlaStrSet) =>
+      // the set of all strings is SetT1(StrT1)
+      mkConst(ExactRef(ex.ID), SetT1(StrT1()))
+
+    //**************************************** EMPTY SETS AND SEQUENCES ***********************************************
+    case OperEx(TypingOper.emptySet, ValEx(TlaStr(elemTypeText))) =>
+      try {
+        val elemType = type1Parser(elemTypeText)
+        mkConst(ExactRef(ex.ID), SetT1(elemType))
+      } catch {
+        case e: Type1ParseError =>
+          throw new TypingInputException(s"Parser error in Typing!EmptySet($elemTypeText): ${e.msg}")
+      }
+
+    case OperEx(TypingOper.emptySeq, ValEx(TlaStr(elemTypeText))) =>
+      try {
+        val elemType = type1Parser(elemTypeText)
+        mkConst(ExactRef(ex.ID), SeqT1(elemType))
+      } catch {
+        case e: Type1ParseError =>
+          throw new TypingInputException(s"Parser error in Typing!EmptySeq($elemTypeText): ${e.msg}")
+      }
+
+    //******************************************** GENERAL OPERATORS **************************************************
+    case OperEx(op, args @ _*) if op == TlaOper.eq || op == TlaOper.ne =>
+      // x = y, x /= y
+      val opsig = OperT1(Seq(VarT1("a"), VarT1("a")), BoolT1())
+      mkApp(ex.ID, opsig, args)
+
+    case OperEx(TlaOper.apply, NameEx(name), args @ _*) =>
+      // F(e_1, ..., e_n)
+      mkAppByName(ExactRef(ex.ID), name, args.map(this(_)) :_*)
+
+    case OperEx(TlaOper.apply, opName, args @ _*) =>
```

Comment:
How would this arise?

Reply:
That is usually a call to a user-defined operator. We call them by names. In the past, we were passing the declaration directly, but that was quite inconvenient for multiple preprocessing and analyses phases.